File: easy-solution/galaxy4276/69.py
class Solution:
  def mySqrt(self, x: int) -> int:
    low, high = 1, x
    mid = (low + high) // 2
    mid_square = mid * mid
    while not (mid_square <= x and (mid + 1) * (mid + 1) > x):
      if mid_square > x:
        high = mid
      else:
        low = mid
      mid = (low + high) // 2
      mid_square = mid * mid

    return mid

File: easy-solution/Antidote21/69.java
class Solution {
    public int mySqrt(int x) {        
        
        if (x == 0 || x == 1) return x;
        
        // Binary Search
        int left = 0, right = x;
        while (left < right) {
            // mid = (left + right) / 2 can overflow if right > Integer.MAX_VALUE - left
            int mid = left + (right - left) / 2;
            
            // same thing here , mid * mid > x can overflow. replace by mid > x / mid
            if (mid > x / mid) {
                right = mid - 1; 
            } else {
                left = mid + 1;
                // if mid * mid < x but (mid + 1) * (mid + 1) > x then mid was the right answer
                if (left > x / left) {
                    return mid;
                }                
            }
        }
        
        return left;
    }
}
